analytical_engine/apps/assortativity/attribute/attribute_assortativity.h

```diff
@@ -0,0 +1,227 @@
+/** Copyright 2020 Alibaba Group Holding Limited.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
+Author: Ning Xin
+*/
+
+#ifndef ANALYTICAL_ENGINE_APPS_ASSORTATIVITY_ATTRIBUTE_ATTRIBUTE_ASSORTATIVITY_H_
+#define ANALYTICAL_ENGINE_APPS_ASSORTATIVITY_ATTRIBUTE_ATTRIBUTE_ASSORTATIVITY_H_
+
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+#include "grape/grape.h"
+
+#include "apps/assortativity/attribute/attribute_assortativity_context.h"
+#include "core/app/app_base.h"
+#include "core/worker/default_worker.h"
+
+namespace gs {
+/**
+ * @brief Compute the attribute assortativity for graph.
+ *  Assortativity measures the similarity of connections in the graph with
+ * respect to the attribute.
+ * @tparam FRAG_T
+ */
+template <typename FRAG_T>
+class AttributeAssortativity
+    : public AppBase<FRAG_T, AttributeAssortativityContext<FRAG_T>>,
+      public grape::Communicator {
+ public:
+  INSTALL_DEFAULT_WORKER(AttributeAssortativity<FRAG_T>,
+                         AttributeAssortativityContext<FRAG_T>, FRAG_T)
+  static constexpr grape::MessageStrategy message_strategy =
+      grape::MessageStrategy::kAlongEdgeToOuterVertex;
+  static constexpr grape::LoadStrategy load_strategy =
+      grape::LoadStrategy::kBothOutIn;
+  using vertex_t = typename fragment_t::vertex_t;
+  using vid_t = typename fragment_t::vid_t;
+  using oid_t = typename fragment_t::oid_t;
+  using vdata_t = typename fragment_t::vdata_t;
+
+  void PEval(const fragment_t& frag, context_t& ctx,
+             message_manager_t& messages) {
+    auto inner_vertices = frag.InnerVertices();
+    for (auto v : inner_vertices) {
+      processVertex(v, frag, ctx, messages);
+    }
+    messages.ForceContinue();
+  }
+
+  void IncEval(const fragment_t& frag, context_t& ctx,
+               message_manager_t& messages) {
+    if (!ctx.merge_stage) {
+      vdata_t source_data;
+      vertex_t u;
+      while (messages.GetMessage(frag, u, source_data)) {
+        vdata_t target_data = frag.GetData(u);
+        attributeMixingCount(source_data, target_data, ctx);
+      }
+      ctx.merge_stage = true;
+      // send message to work 0
```

Comment:
work->worker

Reply:
the mixing matrix of attribute asssortativity app is same with numeric assortativity, but their formula for getting result is different.